mix: make shareable and scan shared file & Update Gradle/Dependencies & Logger & Database Downloading conditions & etc ...

asandikci opened the pull request:
> [!NOTE]
> You can install latest **debug** apk from Actions tab or better **release** apk from our [development build repo](https://github.com/MaintainTeam-Nightly/Hypatia-dev). Also [Nightly F-Droid Repo](https://maintainteam-nightly.github.io/fdroid-pages/fdroid/repo/) can be used for keep in track with latest changes in nightly builds 


---

Note: wiki links not ready yet

## 3.15 Release Notes
### Enduser related
- New Database urls (actually related with previous PR)
- Added disclaimer
- feat: added [share scan feature](WIKI). Now you can share any file to scan with Hypatia !
- ~~feat(WIP): added [File/Folder selection scan feature](WIKI)~~ 
  - will be a separate pr 
- enhancement: [Database downloading conditions](WIKI) updated and fixed unnecessary or missing updates
- enhancement: More modern notification UI
- enhancement: Warn if database signing key is abandoned (actually related with previous PR)
- enhancement: Warn if signing key is wrong/not matches

#### HypatiaDatabases:
- new CIs for creating signature databases and new [signing mechanism](WIKI) (self-hosted)

### Dev Release Notes
- update gradle to 8.9 and java to 21 (for ci)
- drop Android 4 support
- support Android 14
- new [logger mechanism](WIKI) for better debugging and logging
- new AppCompat implementations and AppCompat theme changes
- Actionbar implementation changed to Toolbar implementation
- Cache scan system for shared files that does not have a system path. This is also must2use method for Android 10+ because of some restrictions.
- implemented Application for getting app context rightly and ensure mainactivity has been started
- Commenting to some functions to better understanding 
- added android-env nix file to build on nixos



---

<details><summary>old todos</summary>

READY TO TEST
- [x] update gradle and dependencies
- [x] Solve dependency verification problem
- [x] use AppCompatActivity
- [x] implement single file scan
- [x] solve logger problems.
- [x] fix: toolbar freezing when open (couldn't find any problem, its probably because of debug build)
- [ ] ~~database conditions still needs to be updated~~
- [ ] ~~new database loading/verifying issue~~
  - couldn't reproduce

</details>

asandikci committed:
Update Database Links (again) (#36)

* change cloudflare links to github links
*change default link to codeberg
* warn users if using old database urls
* also warn right after update

asandikci committed:
Merge branch 'dev' into share-file-to-scan

asandikci committed:
update license years and initialize sharescanner

asandikci committed:
add share intent to manifest

asandikci committed:
WIP: share implementation

asandikci committed:
WIP fix manifest

asandikci committed:
WIP

asandikci committed:
WIP

asandikci committed:
WIP

asandikci committed:
update gradle

asandikci committed:
update java to 21

asandikci committed:
change checksum

asandikci commented:
I was just started to implement file sharing thing but updating gradle was blocker ... So pr converted to upgrading gradle pr lol

asandikci committed:
WIP

asandikci committed:
disable checks for now

asandikci commented:
> [!WARNING]
> Disabled dependency verification for now, it should be handled properly later !

edit: solved

asandikci committed:
WIP TEST

asandikci committed:
update sdk versions

* drop android 4 support better maintenance

asandikci committed:
appcompat and theme

asandikci committed:
AppCompat & Toolbar Setup

asandikci committed:
WIP

asandikci committed:
Implement a new Logger

asandikci committed:
add nix file to use android sdk

asandikci committed:
fix: notification panel buttons not workin

asandikci committed:
WIP

asandikci committed:
throw warning if signing key differ

asandikci committed:
database download conditions update

- make logs selectable again
- database.java refactoring
- logger update for debug

asandikci committed:
database conditions update and add disclaimer

asandikci commented:
- [ ] database conditions still needs to be updated

asandikci committed:
fix: hopefully verification metadata

asandikci committed:
add aapt2 verification metadata

- used a debian docker to generate aapt2 verification metadata since my daily nixos build has a modified aapt2

asandikci committed:
warn if file not deleted because of share scan

asandikci commented:
You can test with latest CI output: https://github.com/MaintainTeam/Hypatia/actions/runs/13727619325?pr=38

asandikci committed:
build.gradle file improvements for builds

asandikci commented:
> [!NOTE]
> You can install latest **debug** apk from Actions tab or better **release** apk from our [development build repo](https://github.com/MaintainTeam-Nightly/Hypatia-dev). Also [Nightly F-Droid Repo](https://maintainteam-nightly.github.io/fdroid-pages/fdroid/repo/) can be used for keep in track with latest changes in nightly builds

asandikci commented:
- [x] Found a bug. When realtime scanner enabled share scans will be complated in seconds (because databases already loaded). This results no output to user interface. Only scan start info/notification occurs. This thould be solved !!!
  - edit: temporary fix: calling mainactivity before share scan. But this results a wrong log order, still need some polishment 
- [x] `Since this was a share scan you should manually Remove or Quarantine the file !!!` warning even if no matches found (while cacheScan)
- [x] ... any other bug?
- [ ] merge to dev

Inhishonor commented:
> * `Since this was a share scan you should manually Remove or Quarantine the file !!!` warning even if no matches found (while cacheScan)
>
I can confirm that this is indeed happening. Other than that, this pr looks good.
The only other minor thing I reviewed is related to the capitalization and punctuation of the strings.

Inhishonor commented on app/src/main/res/values/strings.xml:
```diff
@@ -83,4 +84,5 @@
     <string name="lblLinkScannerToggle">Link Scanner</string>
     <string name="confirm_link_scanner_title">Enable link scanner?</string>
     <string name="confirm_link_scanner_summary">[EXPERIMENTAL]\nThis will download an additional database of domains.\nAny domains found in screen content will be checked against the database.\nThis requires granting accessibility service permission manually.\nThis can have a substantial performance impact on screens with large amounts of text.</string>
+    <string name="cache_scan_result_deletion_warning">Since this was a share scan you should manually Remove or Quarantine the file !!!</string>
```
This should be changed to `Since this was a share scan you should manually remove or quarantine the file!`

Inhishonor commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
-    <string name="app_copyright">Copyright 2017-2024 Divested Computing Group, 2025 MaintainTeam</string>
+    <string name="app_disclaimer">This is only a malware scanner which compares files with previously found hashes. Not a magic wand that keeps your entire android super secure. You should be careful what you download/click !!!</string>
```
Similar thing here. `You should be careful what you download/click!`

Inhishonor commented:
One other thing, when I try to enable `Realtime scanning` it crashes the app. This doesn't occur on Version 3.14 from IzzyOnDroid. If you want I can try to debug/fix this crash?

asandikci committed:
improve wording & fix unnecessary warning

asandikci commented:
> One other thing, when I try to enable `Realtime scanning` it crashes the app. This doesn't occur on Version 3.14 from IzzyOnDroid. If you want I can try to debug/fix this crash?

@Inhishonor I couldn't reproduce. Could you please share logs? And I'll be pleasure if you can try to debug/fix :) thanks in advance

asandikci committed:
fix for auto nightly builds

asandikci committed:
fix(tmp): call MainActivity before share scan

- to ensure logs are printed out correctly
- just a temporary fix for now. Log order is not right and HypatiaLogger logQueue is not working as expected

asandikci commented:
HypatiaLogger still needs some polishment but at least working without any fatal error.
@Inhishonor I couldn't see any reason to Realtime scanning to crash. Logs will be really helpful 🤗

Inhishonor commented:
So sorry, I got distracted with other projects. I will look into this shortly @asandikci.

Inhishonor commented:
The problem ended up being that foreground service was not specifying a type which is required in SDK 34. I submitted #44, which should solve this problem.

Inhishonor committed:
Fix Realtime Scanner crashing (#44)

Specified which foregroundServiceType, in order to prevent a crash on newer Android versions.

asandikci commented:
Since commits in this PR are very messy _Squashing_ and merging to dev branch

Inhishonor reviewed (commented):
Looks great! Thanks for implementing this!
